libclc/amdgcn-amdhsa/libspirv/misc/sub_group_shuffle.cl

```diff
@@ -0,0 +1,571 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <spirv/spirv.h>
+
+#define SELF __builtin_amdgcn_mbcnt_hi(-1, __builtin_amdgcn_mbcnt_lo(-1, 0))
+#define SUBGROUP_SIZE __spirv_SubgroupMaxSize()
+
+// Shuffle
+// int __spirv_SubgroupShuffleINTEL<int>(int, unsigned int)
+_CLC_DEF int
+_Z28__spirv_SubgroupShuffleINTELIiET_S0_j(int Data, unsigned int InvocationId) {
+  int self = SELF;
+  int index = InvocationId + (self & ~(SUBGROUP_SIZE - 1));
+  return __builtin_amdgcn_ds_bpermute(index << 2, Data);
+}
+
+// unsigned int __spirv_SubgroupShuffleINTEL<unsigned int>(unsigned int,
+//                                                         unsigned int);
+_CLC_DEF unsigned int
+_Z28__spirv_SubgroupShuffleINTELIjET_S0_j(unsigned int Data,
+                                          unsigned int InvocationId) {
+  return as_uint(
+      _Z28__spirv_SubgroupShuffleINTELIiET_S0_j(as_int(Data), InvocationId));
+}
+// Sub 32-bit types.
```

Comment:
I think it's probably make sense to add similar macro functions for 32-bit types (`int`, `unsigned int` and `float`) and 64-bit types (`long`, `unsigned long` and `double`) to avoid code duplication.

Reply:
That's a good point, done in: d6ffcf37586eeebda5e406d92b87ca0cc975d70c.

Looks like there could be another layer of macros (split per number of arguments), but I worry it would make this file too cryptic.